signer/src/storage/postgres.rs

```diff
@@ -12,6 +52,111 @@ impl PgStore {
     pub async fn connect(url: &str) -> Result<Self, sqlx::Error> {
         Ok(Self(sqlx::PgPool::connect(url).await?))
     }
+
+    /// Write parts of the Stacks Nakamoto block headers to the database.
+    async fn write_stacks_block_header(&self, blocks: &[NakamotoBlock]) -> Result<(), Error> {
+        let block_summaries: Vec<StacksBlockSummary> = blocks
+            .iter()
+            .map(|block| StacksBlockSummary {
+                block_id: block.block_id().to_hex(),
+                chain_length: block.header.chain_length as i64,
+                parent_block_id: block.header.parent_block_id.to_hex(),
+            })
+            .collect();
+
+        if block_summaries.is_empty() {
+            return Ok(());
+        }
+
+        sqlx::query(
+            r#"
+            INSERT INTO sbtc_signer.stacks_blocks (block_hash, block_height, parent_hash, created_at)
+            SELECT
+                decode(block_id, 'hex')
```

Comment:
Why do we both convert the block ID to hex and then later decode it here? Can't we just write the bytes directly?

Reply:
It might be possible, but I couldn't get it to work by going directly to `BYTEA` from a `[u8; N]` once I serialized to JSONB. Let me try one more way; I'm going to try deserializing to `int[]` and then convert that to a `BYTEA`.

Reply:
Right, though couldn't we let the `block_id` type here just be a `Vec<u8>` to keep the type aligned with what the database expects?

Reply:
I'm not sure if the array versus vec difference matters. I'll try though.

Reply:
Yeah I couldn't get it to work in the way that I like. We can send `Vec<u8>` but then we need to do something complicated to convert that to a `BYTEA`. Here's an example of a query that will work but don't like:
```sql
INSERT INTO sbtc_signer.transactions (txid, tx, tx_type, created_at)
SELECT
       decode((SELECT string_agg(lpad(to_hex(x), 2, '0'), '') FROM unnest(txid) AS x), 'hex')
     , decode((SELECT string_agg(lpad(to_hex(x), 2, '0'), '') FROM unnest(tx) AS x), 'hex')
     , tx_type::sbtc_signer.transaction_type
     , CURRENT_TIMESTAMP
FROM JSONB_TO_RECORDSET($1::JSONB) AS x(
      txid      int[]
    , tx        int[]
    , tx_type   VARCHAR
    )
ON CONFLICT DO NOTHING
```
This looks unnecessarily complicated to me and I don't think I can do better if we're starting from an `int[]`.

Reply:
Wait, I'm not following here. I thought we could just dump a `Vec<u8>` into the database as a `BYTEA` type. That's what we already do for bitcoin blocks - which just looks like
```
        sqlx::query!(
            "INSERT INTO sbtc_signer.bitcoin_blocks VALUES ($1, $2, $3, $4, $5)",
            block.block_hash,
            block.block_height,
            block.parent_hash,
            block.confirms,
            block.created_at
        )
```

where `block_hash` is a Vec<u8> which gets written as a BYTEA here. This is tested in the `should_be_able_to_query_bitcoin_blocks` so I'm pretty sure it works.

Reply:
It works for when you directly encode the parameter, but not when you go through `JSONB_TO_RECORDSET`. So the function with the above query works (and `stacks_block_exists` works) but the other queries do not.

Reply:
Right, because you encode the records as a single JSON array before adding it to the query. Which I guess is because you want to write all transactions as a single query as opposed to just writing them one by one? I'd expect the latter to be cleaner but the former to be more efficient - though using the latter with a transaction might be the most efficient option.